pkg/controller/client_builder_dynamic.go

```diff
@@ -0,0 +1,273 @@
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	// Create the namespace if we can't verify it exists.
+	// Tolerate errors, since we don't know whether this component has namespace creation permissions.
+	if _, err := t.CoreClient.Namespaces().Get(t.Namespace, metav1.GetOptions{}); apierrors.IsNotFound(err) {
+		if _, err = t.CoreClient.Namespaces().Create(&v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: t.Namespace}}); err != nil {
+			klog.Warningf("create non-exist namespace %s failed:%v", t.Namespace, err)
+		}
+	}
+
+	// Create the service account
+	sa, err = t.CoreClient.ServiceAccounts(t.Namespace).Create(&v1.ServiceAccount{ObjectMeta: metav1.ObjectMeta{Namespace: t.Namespace, Name: name}})
+	if apierrors.IsAlreadyExists(err) {
+		// If we're racing to init and someone else already created it, re-fetch
+		return t.CoreClient.ServiceAccounts(t.Namespace).Get(name, metav1.GetOptions{})
+	}
+	return sa, err
+}
+
+func (t *DynamicControllerClientBuilder) get(key string) (func(http.RoundTripper) http.RoundTripper, bool) {
+	t.mutex.RLock()
+	defer t.mutex.RUnlock()
+	rt, ok := t.roundTripperFuncMap[key]
+	return rt, ok
+}
+
+func (t *DynamicControllerClientBuilder) set(key string, rt func(http.RoundTripper) http.RoundTripper) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
+	t.roundTripperFuncMap[key] = rt
+}
+
+type tokenSourceImpl struct {
+	clientConfig       *restclient.Config
+	namespace          string
+	serviceAccountName string
+	coreClient         v1core.CoreV1Interface
+	authClient         v1authentication.AuthenticationV1Interface
+}
+
+func (ts *tokenSourceImpl) Token() (*oauth2.Token, error) {
+	tr, err := ts.getToken()
+	if err != nil {
+		return nil, fmt.Errorf("token for %s/%s expired and refresh failed: %v", ts.namespace, ts.serviceAccountName, err)
+	}
+
+	return &oauth2.Token{
+		AccessToken: tr.Status.Token,
+		TokenType:   "Bearer",
+		Expiry:      tr.Status.ExpirationTimestamp.Time,
+	}, nil
+}
+
+func (ts *tokenSourceImpl) getToken() (*v1authenticationapi.TokenRequest, error) {
+	var ret *v1authenticationapi.TokenRequest
+	if err := wait.Poll(100*time.Millisecond, 5*time.Second, func() (bool, error) {
+		tr, inErr := ts.coreClient.ServiceAccounts(ts.namespace).CreateToken(ts.serviceAccountName, &v1authenticationapi.TokenRequest{
+			Spec: v1authenticationapi.TokenRequestSpec{
+				ExpirationSeconds: getInt64Point(defaultExpiration),
+			},
+		})
+		if inErr != nil {
+			return false, fmt.Errorf("get token failed:%v", inErr)
+		}
+
+		valid, inErr := ts.authenticate(tr.Status.Token)
+		if inErr != nil {
+			return false, fmt.Errorf("token authentication failed:%v", inErr)
+		}
+		if !valid {
+			return false, fmt.Errorf("token is invalid: invalid token")
+		}
+		ret = tr
+		return true, nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed to get token for %s/%s : %v", ts.namespace, ts.serviceAccountName, err)
+	}
+
+	return ret, nil
+}
+
+func (ts *tokenSourceImpl) authenticate(token string) (bool, error) {
```

Comment:
Is this necessary? Why is this better than just getting unauthorized?

Reply:
You are right, removed